### PR TITLE
Implement unsolicited message handling while IDLEing

### PR DESCRIPTION
**Pull request by norcalli**
Adapted from https://github.com/jonhoo/rust-imap/pull/114

---
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/rust-imap/130)

### CODE REVIEW

**Comment by norcalli** [comment]
Fixes #128

**Comment by norcalli** [comment]
I had to change around the logic for the "read_response" after "terminate" because there could've been multiple messages sent from the server, so we need to iterate over all of them, but I also need to know when to stop, and the only cue to stop is the `OK` we get from the server for the IDLE `DONE`.

**Comment by norcalli** [comment]
I'm also not sure about making the unsolicited responses be empty by default, because I found it surprising. However, from an engineering perspective, it makes sense because if someone doesn't handle them then they build up in memory. This could be solved by a bounded channel with a sane default, as well, to impose an upper limit.

**Comment by norcalli** [comment]
The test failure appears to be unrelated to any code I wrote.

**Comment by norcalli** [comment]
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `[97, 49, 56]`,
 right: `[97, 49, 57]`', /home/ashkan/works/3rd/rust-imap/src/client.rs:1206:25
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

had an error related to match being one shorter than the client expected. I'm wondering if me doing the matching for the `ok` response in the handle is breaking some invariants.

**Comment by norcalli** [comment]
FYI, I've been running this for the past week without any crashes or problems. I'm using it as my main ingestion pipeline for my mail now.

**Comment by jonhoo** [comment]
I've been away on vacation for the past few weeks, so haven't had a chance to look. Will try to take a look once I get the time! Thanks for writing this up! :D

File: src/parse.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ where
                     match map(resp)? {
                         MapOrNot::Map(t) => things.push(t),
                         MapOrNot::Not(resp) => match handle_unilateral(resp, unsolicited) {
-                            Some(Response::Fetch(..)) => continue,
+                            // Some(Response::Fetch(..)) => continue,
```
Comment (jonhoo) [inline src/parse.rs]: Why this change?

File: src/extensions/idle.rs [hunk 2]
```diff
@@ -77,31 +84,53 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
...
+                        || e.kind() == io::ErrorKind::WouldBlock =>
+                {
+                    if self.keepalive.is_some() {
+                        // we need to refresh the IDLE connection
+                        self.terminate(true)?;
+                        self.init()?;
+                    }
+                }
+                Err(err) => return Err(err),
+                Ok(_) => {
```
Comment (jonhoo) [inline src/extensions/idle.rs]: I don't understand this clause at all. Why do we need this nested loop construct? And why do we need to terminate the IDLE for each iteration?

File: src/parse.rs [hunk 3]
```diff
@@ -202,7 +199,7 @@ pub fn parse_mailbox(
                 if let imap_proto::Status::Ok = status {
                 } else {
                     // how can this happen for a Response::Data?
-                    unreachable!();
+                    unreachable!("Received something other than OK from mailbox data");
```
Comment (jonhoo) [inline src/parse.rs]: This should probably include `status` for convenience.

File: src/parse.rs [hunk 4]
```diff
@@ -297,29 +292,124 @@ pub fn parse_ids(
...
+                    ..
+                },
+            )) => {
+                match status {
+                    imap_proto::Status::Ok => return Ok(true),
+                    imap_proto::Status::Bad => {
+                        return Err(Error::Bad(information.unwrap_or("").to_owned()))
+                    }
+                    imap_proto::Status::No => {
+                        return Err(Error::Bad(information.unwrap_or("").to_owned()))
```
Comment (jonhoo) [inline src/parse.rs]: Don't we also have `Error::No`?

File: src/extensions/idle.rs [hunk 5]
```diff
@@ -77,31 +84,53 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
...
+                    let _ = parse::parse_idle(&buffer, &mut self.unsolicited_responses_tx)?;
```
Comment (jonhoo) [inline src/extensions/idle.rs]: Specifically, shouldn't this use the return value from `parse_idle` to know whether or not `DONE` was sent?

File: src/parse.rs [hunk 6]
```diff
@@ -297,29 +292,124 @@ pub fn parse_ids(
...
+                    }
+                    _ => unreachable!(),
+                }
+            }
+            Ok((rest, data)) => {
+                lines = rest;
+                if let Some(resp) = handle_unilateral(data, unsolicited) {
```
Comment (jonhoo) [inline src/parse.rs]: We should probably rename `handle_unilateral` to `try_handle_unilateral` to make it clearer that `Some` is returned if it wasn't handled. That's orthogonal to your change though :)

File: src/types/mod.rs [hunk 7]
```diff
@@ -204,6 +205,60 @@ pub use self::capabilities::Capabilities;
...
+    TryCreate,
+    /// Next unique identifier value.
+    UidNext(u32),
+    /// The unique identifier validity value.
+    UidValidity(u32),
+    /// First message without the \Seen flag set.
+    Unseen(u32),
+}
+
+impl<'a> From<imap_proto::types::ResponseCode<'a>> for ResponseCode {
```
Comment (jonhoo) [inline src/types/mod.rs]: I wonder if [`ToOwned`](https://doc.rust-lang.org/std/borrow/trait.ToOwned.html) is the trait to implement here.

File: src/unsolicited_responses.rs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        UnsolicitedResponseSender {
+            sender,
+            allow: EnumSet::empty(),
+        }
+    }
+
+    // Set the new filter mask, and remove unwanted responses from the current queue.
+    pub(crate) fn request(
+        &mut self,
+        receiver: &mpsc::Receiver<UnsolicitedResponse>,
```
Comment (jonhoo) [inline src/unsolicited_responses.rs]: I'm not sure I understand what this receiver is? Where does it come from?

File: src/unsolicited_responses.rs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        mask: EnumSet<UnsolicitedResponseCategory>,
+    ) {
+        self.allow = mask;
+        for message in receiver.try_iter() {
```
Comment (jonhoo) [inline src/unsolicited_responses.rs]: This also seems pretty strange -- what about future messages? Are they never forwarded then?

**Comment by jonhoo** [comment]
Hi @norcalli -- a gentle ping on this :)

**Comment by norcalli** [comment]
@jonhoo hey sorry for the late reply, gonna take a look at this today.

**Comment by hpk42** [comment]
@norcalli just so you i'd be interested in your PR landing but my Rust-Fu is not good enough yet to help much.

**Comment by jonhoo** [comment]
@norcalli Another gentle ping on this :)

File: src/unsolicited_responses.rs (hunk 9 above)
Comment (jonhoo) [reply src/unsolicited_responses.rs:27]: Long time no see :) Did you ever get around to doing this rewrite?

**Comment by jonhoo** [comment]
@norcalli Another gentle ping on this :)

**Comment by jonhoo** [comment]
@norcalli I'd like to try to push out 3.0.0 soon — want to give this another stab?

**Comment by jonhoo** [comment]
Replaced by #186.
